### PR TITLE
SRVLOGIC-289 - Introduce kn-workflow-cli-artifacts image bundle

### PR DESCRIPTION
**Pull request by ricardozanini**
**Description of the change:**
In this PR, we introduce a new image for OpenShift Serverless Logic kn-plugin download. 

A few points notes:

1. The envvar `KN_WORK_DIR` must be immutable and it's not an ARG anymore. We did this since this dir value is reused in the subsequent build chain downstream and it's hardcoded in the descriptor. 
2. The name of the artifacts must have the `kn-workflow-` suffix as you can see in the descriptors

This will work only under RH's VPN:

```
cekit -v --descriptor images/osl-cli.yaml build docker --build-args KN_ARTIFACT_URL=<the URL>
```

@rgdoliveira Can you please validate this image? 

This is how the image looks like:

```Dockerfile

# This is a Dockerfile for the kn-workflow-cli-artifacts:1.33.0-snapshot image.

## START builder image packager:1.33.0-snapshot
## \
    FROM registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161 AS packager
    USER root

###### START module 'com.redhat.osl.cli.packager:1.33.0-snapshot'
###### \
        # Copy 'com.redhat.osl.cli.packager' module general artifacts to '/tmp/artifacts/' destination
        COPY \
            LICENSE \
            /tmp/artifacts/
        # Copy 'com.redhat.osl.cli.packager' module content
        COPY modules/com.redhat.osl.cli.packager /tmp/scripts/com.redhat.osl.cli.packager
        # Switch to 'root' user for package management for 'com.redhat.osl.cli.packager' module defined packages
        USER root
        # Install packages defined in the 'com.redhat.osl.cli.packager' module
        RUN microdnf --setopt=install_weak_deps=0 --setopt=tsflags=nodocs install -y zip tar gzip wget \
            && microdnf clean all \
            && rpm -q zip tar gzip wget
        ARG KN_ARTIFACT_URL="http://download.eng.bos.redhat.com/staging-cds/etera/openshift-serverless-clients/1/1.11/1.11.2-4/signed"

        # Set 'com.redhat.osl.cli.packager' module defined environment variables
        ENV \
            KN_WORK_DIR="/opt/app-root/src/go/src/github.com/knative/client"
        # Custom scripts from 'com.redhat.osl.cli.packager' module
        USER root
        RUN [ "sh", "-x", "/tmp/scripts/com.redhat.osl.cli.packager/install.sh" ]
###### /
###### END module 'com.redhat.osl.cli.packager:1.33.0-snapshot'

###### START image 'packager:1.33.0-snapshot'
###### \
###### /
###### END image 'packager:1.33.0-snapshot'


## /
## END builder image

## START target image kn-workflow-cli-artifacts:1.33.0-snapshot
## \
    FROM registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161


    USER root

###### START module 'com.redhat.osl.cli.artifacts:1.33.0-snapshot'
###### \
        # Copy 'com.redhat.osl.cli.artifacts' module stage artifacts
        COPY --from=packager /opt/app-root/src/go/src/github.com/knative/client/kn-workflow-linux-amd64.tar.gz /usr/share/kn/linux_amd64/kn-workflow-linux-amd64
        COPY --from=packager /opt/app-root/src/go/src/github.com/knative/client/kn-workflow-linux-ppc64le.tar.gz /usr/share/kn/linux_ppc64le/kn-workflow-linux-ppc64le
        COPY --from=packager /opt/app-root/src/go/src/github.com/knative/client/kn-workflow-linux-s390x.tar.gz /usr/share/kn/linux_s390x/kn-workflow-linux-s390x
        COPY --from=packager /opt/app-root/src/go/src/github.com/knative/client/kn-workflow-macos-amd64.tar.gz /usr/share/kn/macos/kn-workflow-macos-amd64
        COPY --from=packager /opt/app-root/src/go/src/github.com/knative/client/kn-workflow-windows-amd64.zip /usr/share/kn/windows/kn-workflow-windows-amd64
        # Copy 'com.redhat.osl.cli.artifacts' module content
        COPY modules/com.redhat.osl.cli.artifacts /tmp/scripts/com.redhat.osl.cli.artifacts
        # Custom scripts from 'com.redhat.osl.cli.artifacts' module
        USER root
        RUN [ "sh", "-x", "/tmp/scripts/com.redhat.osl.cli.artifacts/install.sh" ]
###### /
###### END module 'com.redhat.osl.cli.artifacts:1.33.0-snapshot'

###### START image 'kn-workflow-cli-artifacts:1.33.0-snapshot'
###### \
        # Set 'kn-workflow-cli-artifacts' image defined labels
        LABEL \
            com.redhat.component="openshift-serverless-1-kn-workflow-cli-artifacts-rhel8-container" \
            description="'kn-workflow' CLI cross platform binaries for managing Red Hat OpenShift Serverless Logic 1" \
            io.cekit.version="4.11.0" \
            io.k8s.description="'kn-workflow' CLI cross platform binaries for managing Red Hat OpenShift Serverless Logic 1" \
            io.k8s.display-name="Red Hat OpenShift Serverless Logic 1 kn-workflow CLI artifacts" \
            io.openshift.build.source-location="https://github.com/apache/incubator-kie-tools/" \
            maintainer="serverless-support@redhat.com" \
            name="openshift-serverless-1/kn-workflow-cli-artifacts-rhel8" \
            summary="Red Hat OpenShift Serverless Logic 1 kn-workflow CLI artifacts" \
            version="1.33.0-snapshot"
###### /
###### END image 'kn-workflow-cli-artifacts:1.33.0-snapshot'



    # Switch to 'root' user and remove artifacts and modules
    USER root
    RUN rm -rf "/tmp/scripts" "/tmp/artifacts"
    # Define the user
    USER 65532
## /
## END target image
```

**Motivation for the change:**
See https://issues.redhat.com/browse/SRVLOGIC-289

**Checklist**

- [ ] Add or Modify a unit test for your change
- [ ] Have you verified that tall the tests are passing?

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

File: images/osl-cli.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+- name: "packager"
+  version: 1.33.0-snapshot
+  from: "registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161"
```
Comment (rgdoliveira) [inline images/osl-cli.yaml]:
```suggestion
  from: "registry.access.redhat.com/ubi8/ubi-minimal:latest"
```

File: images/osl-cli.yaml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  modules:
+    repositories:
+      - path: modules
+    install:
+      - name: com.redhat.osl.cli.packager
+
+- name: "kn-workflow-cli-artifacts"
+  version: 1.33.0-snapshot
+  description: "Red Hat OpenShift Serverless Logic 1 kn-workflow CLI artifacts"
+  from: "registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161"
```
Comment (rgdoliveira) [inline images/osl-cli.yaml]:
```suggestion
  from: "registry.access.redhat.com/ubi8/ubi-minimal:latest"
```

File: images/osl-cli.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      value: "https://github.com/apache/incubator-kie-tools/"
+    - name: maintainer
+      value: "serverless-support@redhat.com"
+  modules:
+    repositories:
+      - path: modules
+    install:
+      - name: com.redhat.osl.cli.artifacts
+  run:
+    user: 65532
```
Comment (rgdoliveira) [inline images/osl-cli.yaml:53]: We need a block like this to build in OSBS:

```
packages:
  manager: microdnf
  content_sets_file: content_sets.yaml

osbs:
  configuration:
    container:
      platforms:
        only:
        - x86_64
        - aarch64
        - ppc64le
      compose:
        pulp_repos: true
  repository:
    name: containers/openshift-serverless-1-logic-kn-workflow-cli-artifacts
    branch: openshift-serverless-1.33-rhel-8
```

Note that the contents_set.yaml was removed in upstream but I will recreate it in midstream as part of SRVLOGIC-287

File: images/osl-cli.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+- name: "packager"
+  version: 1.33.0-snapshot
+  from: "registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161"
+  modules:
+    repositories:
+      - path: modules
+    install:
+      - name: com.redhat.osl.cli.packager
+
+- name: "kn-workflow-cli-artifacts"
```
Comment (rgdoliveira) [inline images/osl-cli.yaml]:
```suggestion
- name: "openshift-serverless-1/logic-kn-workflow-cli-artifacts-rhel8"
```

File: images/osl-cli.yaml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  version: 1.33.0-snapshot
+  description: "Red Hat OpenShift Serverless Logic 1 kn-workflow CLI artifacts"
+  from: "registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161"
+  labels:
+    - name: com.redhat.component
+      value: "openshift-serverless-1-kn-workflow-cli-artifacts-rhel8-container"
```
Comment (rgdoliveira) [inline images/osl-cli.yaml]:
```suggestion
      value: "openshift-serverless-1-logic-kn-workflow-cli-artifacts-rhel8-container"
```

File: images/osl-cli.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    - name: name
+      value: "openshift-serverless-1/kn-workflow-cli-artifacts-rhel8"
+    - name: version
+      value: "1.33.0-snapshot"
```
Comment (rgdoliveira) [inline images/osl-cli.yaml:34]: I dont think we need this "name" and "version" block here, we dont have it for other images. I would drop it to make it simpler.

**Review by rgdoliveira — COMMENTED** [review]
I've made some comments

File: images/modules/com.redhat.osl.cli.packager/module.yaml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    - wget
+envs:
+  - name: KN_WORK_DIR
+    value: "/opt/app-root/src/go/src/github.com/knative/client"
+    description: "Immutable working directory for this image, please don't change it"
+args:
+  - name: KN_ARTIFACT_URL
+    description: "Base URL from where to download the artifacts"
+    value: "http://download.eng.bos.redhat.com/staging-cds/etera/openshift-serverless-clients/1/1.11/1.11.2-4/signed"
+    example: "http://download.eng.bos.redhat.com/staging-cds/etera/openshift-serverless-clients/1/1.11/1.11.2-4/signed"
```
Comment (dsimansk) [inline images/modules/com.redhat.osl.cli.packager/module.yaml]: Not sure about the current policy, if we can use internal URLs in public repos?

Comment (mareknovotny) [reply images/modules/com.redhat.osl.cli.packager/module.yaml]: these links or refs should be externalized through properties.

File: images/modules/com.redhat.osl.cli.packager/install.sh [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+set -e
+
+mkdir -p "${KN_WORK_DIR}"
+cd "${KN_WORK_DIR}"
+
+mv  "/tmp/LICENSE" .
+
+wget -P "${KN_WORK_DIR}" "${KN_ARTIFACT_URL}/amd64/linux/kn-workflow-linux-amd64"
+wget -P "${KN_WORK_DIR}" "${KN_ARTIFACT_URL}/ppc64le/linux/kn-workflow-linux-ppc64le"
+wget -P "${KN_WORK_DIR}" "${KN_ARTIFACT_URL}/s390x/linux/kn-workflow-linux-s390x"
```
Comment (rgdoliveira) [inline images/modules/com.redhat.osl.cli.packager/install.sh:25]: I dont think we build images for s390x (system Z), but we do build images for arm (aarch64), shouldn't it be replaced

File: images/osl-cli.yaml (hunk 3 above)
Comment (ricardozanini) [inline images/osl-cli.yaml:53]: Can you please add a suggestion so we can include everything at once? @rgdoliveira

File: images/osl-cli.yaml (hunk 6 above)
Comment (ricardozanini) [reply images/osl-cli.yaml:34]: Version is required.

File: images/modules/com.redhat.osl.cli.packager/module.yaml (hunk 7 above)
Comment (ricardozanini) [reply images/modules/com.redhat.osl.cli.packager/module.yaml]: These links are externalized, this is just a default value. There's a `build-args` argument as I highlighted in the PR description.

File: images/modules/com.redhat.osl.cli.packager/install.sh (hunk 8 above)
Comment (ricardozanini) [reply images/modules/com.redhat.osl.cli.packager/install.sh:25]: this is not the image, but the client. If we build the kn client for these arch so then we must do the same for the plugin.

File: images/osl-cli.yaml (hunk 3 above)
Comment (rgdoliveira) [reply images/osl-cli.yaml:53]: ok, I will delete this comment and add it as a suggestion

Comment (rgdoliveira) [inline images/osl-cli.yaml:53]:
```suggestion
    user: 65532
    
packages:
  manager: microdnf
  content_sets_file: content_sets.yaml

osbs:
  configuration:
    container:
      platforms:
        only:
        - x86_64
        - aarch64
        - ppc64le
      compose:
        pulp_repos: true
  repository:
    name: containers/openshift-serverless-1-logic-kn-workflow-cli-artifacts
    branch: openshift-serverless-1.33-rhel-8    
```

File: images/osl-cli.yaml [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+- name: "kn-workflow-cli-artifacts"
+  version: 1.33.0-snapshot
+  description: "Red Hat OpenShift Serverless Logic 1 kn-workflow CLI artifacts"
+  from: "registry.access.redhat.com/ubi8/ubi-minimal:8.9-1161"
+  labels:
+    - name: com.redhat.component
+      value: "openshift-serverless-1-kn-workflow-cli-artifacts-rhel8-container"
+    - name: name
+      value: "openshift-serverless-1/kn-workflow-cli-artifacts-rhel8"
```
Comment (rgdoliveira) [inline images/osl-cli.yaml:32]:
```suggestion

```

File: images/osl-cli.yaml (hunk 6 above)
Comment (rgdoliveira) [reply images/osl-cli.yaml:34]: so maybe drop the name, I added a suggestion.

File: images/modules/com.redhat.osl.cli.packager/module.yaml (hunk 7 above)
Comment (ricardozanini) [reply images/modules/com.redhat.osl.cli.packager/module.yaml]: I'll remove the value, so the build fails if no args are passed. I just did this 'cause the original Dockerfile has it hardcoded.

File: images/modules/com.redhat.osl.cli.packager/module.yaml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    - gzip
+    - wget
+envs:
+  - name: KN_WORK_DIR
+    value: "/opt/app-root/src/go/src/github.com/knative/client"
+    description: "Immutable working directory for this image, please don't change it"
+args:
+  - name: KN_ARTIFACT_URL
+    description: "Base URL from where to download the artifacts"
+    value: "http://download.eng.bos.redhat.com/staging-cds/etera/openshift-serverless-clients/1/1.11/1.11.2-4/signed"
```
Comment (ricardozanini) [inline images/modules/com.redhat.osl.cli.packager/module.yaml]:
```suggestion
```

File: images/osl-cli.yaml (hunk 6 above)
Comment (ricardozanini) [reply images/osl-cli.yaml:34]: @rgdoliveira I'm follwing the same approach of the original image. I think we should keep both.

File: images/osl-cli.yaml (hunk 9 above)
Comment (ricardozanini) [reply images/osl-cli.yaml:32]: I'll keep following the same pattern as the original kn cli image. We can introduce these fields to the other images as well.

**Comment by ricardozanini** [comment]
@mareknovotny @rgdoliveira @dsimansk can you please take a second look?

File: images/osl-cli.yaml (hunk 6 above)
Comment (rgdoliveira) [reply images/osl-cli.yaml:34]: please check the other images we productize, I cant see neither `name` nor `version` there. That's why I'm saying that maybe it is better to remove.

File: images/osl-cli.yaml (hunk 9 above)
Comment (rgdoliveira) [reply images/osl-cli.yaml:32]: ok, as you wish.

File: images/osl-cli.yaml (hunk 6 above)
Comment (rgdoliveira) [reply images/osl-cli.yaml:34]: but anyway, just my opinion, just go with it whatever way you prefer

**Review by mareknovotny — APPROVED** [review]

**Pull request merged**